File: pset6/credit/credit.py
from cs50 import get_string
import math

while True:
    number = get_string("Number: ")
    try:
        int(number)
        break
    except:
        pass


length = len(number)
first_digit = number[0]
first_two_digits = number[:2]


def algorithm(n=number):
    total = 0 # Initializes the sum

    # Iterates over the digits, starting from the end to bc reference is at the second to last number
    for i in range(-1, -length - 1, -1):
        if i % 2 == 0: # If the current digit is every other number starting from the second last
            # The digit is multiplied
            digit = 2*int(n[i])

            # If the result is a two digit number, we split the number
            if digit > 9:
                # And then add the two digits
                total += (math.floor(digit / 10) + digit % 10)
            else:
                # Otherwise, add the digit itself
                total += digit
        else:
            total += int(n[i])
    return total % 10 == 0 # Returns whether or not the sum is a multiple of 10


if (algorithm()):
    if (length == 15 and (first_two_digits == '34' or first_two_digits == '37')):
        print("AMEX\n")
    elif length == 16 and (int(first_two_digits) in range(51, 56)): # Cheks if first two digits b/w 51 and 55
        print("MASTERCARD\n")
    elif (length == 13 or length == 16) and first_digit == '4':
        print("VISA\n")
    else:
        print("INVALID\n")
else:
    print("INVALID\n")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool algorithm(long n);

int main(void)
{
    // Asks for user input
    long n = get_long("Number: ");

    // Finds the length of the number
    char str[256];
    int len = sprintf(str, "%li", n);

    // Finds the first, and first two digits of the number
    long first_digit = n / (long) pow((double) 10, (double) len - 1);
    long first_two_digits = n / (long) pow((double) 10, (double) len - 2);

    // Checks if the card complies with Luhn's algorithm
    if (algorithm(n))
    {
        // Checks for AMEX
        if (len == 15 && (first_two_digits == 34 || first_two_digits == 37))
        {
            printf("AMEX\n");
        }

        // Checks for Mastercard
        else if (len == 16 &&
            (
                first_two_digits == 51 ||
                first_two_digits == 52 ||
                first_two_digits == 53 ||
                first_two_digits == 54 ||
                first_two_digits == 55
            ))
        {
            printf("MASTERCARD\n");
        }
        // Checks for VISA
        else if ((len == 13 || len == 16) && first_digit == 4)
        {
            printf("VISA\n");
        }
        // Otherwise invalid
        else
        {
            printf("INVALID\n");
        }
    }
    // Otherwise invalid
    else
    {
        printf("INVALID\n");
    }

};

bool algorithm(long n)
{
    long tmp = n;

    int sum = 0;
    int counter = 0;

    while (tmp >= 1)
    {
        int current_digit = tmp % 10;
        if (counter % 2 == 1)
        {

            if (2*current_digit > 9)
            {
                sum += ((2*current_digit / 10) + (2*current_digit % 10));
            }
            else
            {
                sum += 2*current_digit;
            }
        }
        else
        {
            sum += current_digit;
        }

        tmp /= 10;
        counter++;
    }

    if (sum % 10 == 0)
    {
        return true;
    }

    return false;
};